File: 559. Maximum Depth of N-ary Tree/main.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> children;

    public Node() {}

    public Node(int _val,List<Node> _children) {
        val = _val;
        children = _children;
    }
};
*/
class Solution {
    public int maxDepth(Node root) {
        
        // DFS
        if(root==null){
            return 0;
        } else if (root.children == null){
            return 1;
        } else {
            int cMax = 0;
            for (Node child:root.children){
                int newValue = maxDepth(child);
                if (newValue > cMax){
                    cMax = newValue;
                }
            }
            return cMax+1;
        }
        
        // BFS
//         if(root == null) return 0;
    
//         Queue<Node> queue = new LinkedList<>();
//         queue.offer(root);

//         int depth = 0;

//         while(!queue.isEmpty())
//         {
//             int size = queue.size();

//             for(int i = 0; i < size; i++)
//             {
//                 Node current = queue.poll();
//                 for(Node child: current.children) queue.offer(child);
//             }

//             depth++;
//         }

//         return depth;
    }
}

File: 559. Maximum Depth of N-ary Tree/main.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val, children):
        self.val = val
        self.children = children
"""
from collections import deque
class Solution:
    def maxDepth(self, root: 'Node') -> int:
        # dfs
        if not root:
            return 0
        elif not root.children:
            return 1
        else:
            depth = 0
            for child in root.children:
                depth = max(depth,self.maxDepth(child))
            return depth+1
        
        # bfs
        # if not root:
        #     return 0
        # else:
        #     dq = deque()
        #     dq.append(root)
        #     depth = 0
        #     while len(dq) >0:
        #         depth+=1
        #         breadth = len(dq)
        #         for _ in range(breadth):
        #             node = dq.popleft()
        #             if node.children:
        #                 for child in node.children:
        #                     dq.append(child)
        #     return depth